lib/http2/debug_state.c

```diff
@@ -0,0 +1,210 @@
+#include "h2o/http2.h"
+#include "h2o/http2_internal.h"
+
+static const char *debug_state_string_open = "OPEN";
+static const char *debug_state_string_half_closed_remote = "HALF_CLOSED_REMOTE";
+static const char *debug_state_string_reserved_local = "RESERVED_LOCAL";
+
+const char *get_debug_state_string(h2o_http2_stream_t *stream)
+{
+    if (h2o_http2_stream_is_push(stream->stream_id)) {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+            return debug_state_string_reserved_local;
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_half_closed_remote;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    } else {
+        switch (stream->state) {
+        case H2O_HTTP2_STREAM_STATE_RECV_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_RECV_BODY:
+        case H2O_HTTP2_STREAM_STATE_REQ_PENDING:
+        case H2O_HTTP2_STREAM_STATE_SEND_HEADERS:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY:
+        case H2O_HTTP2_STREAM_STATE_SEND_BODY_IS_FINAL:
+            return debug_state_string_open;
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+h2o_http2_debug_state_t *h2o_get_http2_debug_state(h2o_req_t *req, int hpack_enabled)
+{
+    /* if the request is sent via HTTP/1, return 404 response */
+    if (req->version < 0x200)
+        return NULL;
+
+    h2o_http2_conn_t *conn = (h2o_http2_conn_t*)req->conn;
+    struct st_h2o_http2_debug_state_t *state = h2o_mem_alloc_pool(&req->pool, sizeof(*state));
+    *state = (struct st_h2o_http2_debug_state_t){{NULL}};
+
+    state->conn_flow_in = conn->_write.window._avail;
+    state->conn_flow_out = conn->_write.window._avail;
+
+    /* calculate total number of JSON lines */
+    size_t nr_resp = 3;
+    nr_resp += conn->num_streams.pull.open + conn->num_streams.push.open; // streams
+    if (hpack_enabled) {
+        nr_resp += 3 + conn->_input_header_table.num_entries + conn->_output_header_table.num_entries; // hpack
+    }
+    h2o_vector_reserve(&req->pool, &state->json, nr_resp);
+
+    char check[1]; /* for only checking the size of output */
+
+#define OUTPUT(_fmt, ...)                                        \
+    do {                                                         \
+        int size = snprintf(check, 1, _fmt, __VA_ARGS__);        \
+        h2o_iovec_t v;                                           \
+        v.base = h2o_mem_alloc_pool(&req->pool, size + 1);             \
+        v.len = snprintf(v.base, size + 1, _fmt, __VA_ARGS__);   \
+        state->json.entries[state->json.size++] = v; \
+    } while (0)
```

Comment:
I'd suggest making this a function (instead of macro), considering the fact that the arguments are evaluated more than once, and because this is not performance-critical.